# Introduction
**This will be your workspace for the [Machine Learning course](https://www.kaggle.com/learn/machine-learning).**

You will need to translate the concepts to work with the data in this notebook, the Iowa data. Each page in the Machine Learning course includes instructions for what code to write at that step in the course.

# Write Your Code Below

In [2]:
import pandas as pd

main_file_path = '../input/house-prices-advanced-regression-techniques/train.csv' # this is the path to the Iowa data that you will use
data = pd.read_csv(main_file_path)

# Run this code block with the control-enter keys on your keyboard. Or click the blue botton on the left
print('Some output from running this cell')

Some output from running this cell


In [3]:
data.describe()

Id   MSSubClass      ...             YrSold      SalePrice
count  1460.000000  1460.000000      ...        1460.000000    1460.000000
mean    730.500000    56.897260      ...        2007.815753  180921.195890
std     421.610009    42.300571      ...           1.328095   79442.502883
min       1.000000    20.000000      ...        2006.000000   34900.000000
25%     365.750000    20.000000      ...        2007.000000  129975.000000
50%     730.500000    50.000000      ...        2008.000000  163000.000000
75%    1095.250000    70.000000      ...        2009.000000  214000.000000
max    1460.000000   190.000000      ...        2010.000000  755000.000000

[8 rows x 38 columns]

In [4]:
data.columns[data.dtypes=='int']

Index(['Id', 'MSSubClass', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF',
       'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea',
       'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr',
       'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageCars',
       'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch',
       'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')

In [5]:
train_features = ['LotArea','YearBuilt','1stFlrSF', '2ndFlrSF','FullBath', 'BedroomAbvGr','TotRmsAbvGrd']
target_features = ['SalePrice']
x = data[train_features]
y = data[target_features]
from sklearn.tree import DecisionTreeRegressor
data_model = DecisionTreeRegressor()
data_model.fit(x,y)
predicted_value = data_model.predict(x)
from sklearn.metrics import mean_absolute_error
error = mean_absolute_error(y,predicted_value)
error

62.354337899543388

In [6]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest = train_test_split(x,y,random_state=0)
data_model = DecisionTreeRegressor()
data_model.fit(xtrain,ytrain)
predicted_value = data_model.predict(xtest)
new_error = mean_absolute_error(ytest,predicted_value)
new_error

32996.257534246579

In [7]:
def get_mae(max_leaf_nodes,xtrain,ytrain,xtest,ytest):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes)
    model.fit(xtrain,ytrain)
    predicted_val = model.predict(xtest)
    error = mean_absolute_error(ytest,predicted_val)
    return error
for max_leaf_nodes in [5,50,500,5000]:
    err = get_mae(max_leaf_nodes,xtrain,ytrain,xtest,ytest)
    print("max_leaf_nodes: %d  \t\t mean absolute error is : %d" %(max_leaf_nodes,err))
    

max_leaf_nodes: 5  		 mean absolute error is : 35190
max_leaf_nodes: 50  		 mean absolute error is : 27825
max_leaf_nodes: 500  		 mean absolute error is : 32473
max_leaf_nodes: 5000  		 mean absolute error is : 33273


### Here we can say the optimum max_leaf_nodes is 50

## Random Forest

In [8]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error
model = RandomForestRegressor()
model.fit(xtrain,ytrain)
pred_val = model.predict(xtest)
err = mean_absolute_error(ytest,pred_val)
err

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  after removing the cwd from sys.path.


24190.010867579909

#### Here we can see the Random forest is working pretty fine(24190) with out tuning parameters,where as the Decision Tree with the tuning parameters(27825)


**If you have any questions or hit any problems, come to the [Learn Discussion](https://www.kaggle.com/learn-forum) for help. **

**Return to [ML Course Index](https://www.kaggle.com/learn/machine-learning)**